processing/src/main/java/org/apache/druid/query/planning/PreJoinableClause.java

```diff
@@ -64,6 +68,29 @@ public DataSource getDataSource()
     return dataSource;
   }
 
+  /**
+   * If the data source is a {@link RestrictedDataSource} with a {@link NoRestrictionPolicy}, unwraps it to a table and return.
+   * <p>
+   * This is a temporary workaround to allow the planner to work with {@link RestrictedDataSource} as the right-side join.
+   */
+  public DataSource maybeUnwrapRestrictedDataSource()
+  {
+    if (dataSource instanceof RestrictedDataSource) {
+      RestrictedDataSource restricted = (RestrictedDataSource) dataSource;
+      if (restricted.getPolicy() instanceof NoRestrictionPolicy) {
+        return restricted.getBase();
+      } else {
+        throw new QueryUnsupportedException(StringUtils.format(
```

Comment:
Should this be a different error like https://github.com/apache/druid/blob/da0b404aa214494c3fa1ed24d8b0b2ff49cac331/processing/src/main/java/org/apache/druid/error/DruidException.java#L166